model/riscv_fdext_regs.sail

```diff
@@ -210,6 +245,67 @@ function wF_bits(i: bits(5), data: flenbits) -> unit = {
 
 overload F = {rF_bits, wF_bits, rF, wF}
 
+val handle_illegal : unit -> unit effect {escape, rreg, wreg}
+
+/* ---- Register Read/ Writes */
+
+val rX_or_rF_32 : (bool, bits(5)) -> (Retired, bits(32)) effect {escape, rreg}
+function rX_or_rF_32(zfinx_en, i) = {
+  let rval = if zfinx_en then nan_unbox(zfinx_en, rX(unsigned(i))) else nan_unbox(zfinx_en, rF(unsigned(i)));
+  (RETIRE_SUCCESS, rval)
+}
+
+val rX_or_rF_64 : (bool, bits(5)) -> (Retired, bits(64)) effect {escape, rreg, wreg}
+function rX_or_rF_64(zfinx_en, i) = {
+  if zfinx_en & sizeof (xlen) == 32 then {
+    if i == 0b00000 then
+      (RETIRE_SUCCESS, sail_zeros(64))
+    else if i[0] == bitzero then (RETIRE_SUCCESS, rX(unsigned(i+1)) @ rX(unsigned(i)))
+    else { handle_illegal(); (RETIRE_FAIL, sail_zeros(64)) }
+  }
+  else if zfinx_en & sizeof (xlen) == 64 then
+    (RETIRE_SUCCESS, rX(unsigned(i)))
+  else if ~(zfinx_en) & sizeof (flen) == 64 then
+    (RETIRE_SUCCESS, rF(unsigned(i)))
+  else {
+    assert(false);
+    (RETIRE_FAIL, sail_zeros(64))
+  }
+}
```

Comment:
So, this function (and the write version, though one argument doesn't apply) is where all the real problems lie.

Firstly, `(RETIRE_FAIL, sail_zeros(64))` should be a huge red flag saying "the types are wrong"; that's how you write it in Verilog, but it's not how you write it in a language with a proper type system, you don't just shove dummy value into things with a separate validity field, you use a proper discriminated union, which means enum in Sail and other modern languages; in this case there's option already in the language that does exactly what you want.

Secondly, the use of RETIRE_SUCCESS/FAIL here is a layering violation. If you look throughout the model, RETIRE_SUCCESS/FAIL are only ever used in places that are exclusively about instruction execution, and within the riscv_insts_* files themselves (ignoring iris, which is a load of files all concatenated together, and has been removed, and riscv_jalr_rmem/seq which are just there so two different implementations of JALR's execute clause can be used). There is never a pattern like this. Functions like this should be free from the semantics of instruction retiring; for example, if someone were to extend the model with a debug module implementation, register reading and writing is something that it can do (via its abstract commands) but there is no notion of retiring there. The correct design is to have this be independent of instruction execution semantics and let the consumers deal with translating that into RETIRE_FAIL on error, which is then consistent with how we deal with things like memory translation errors and extension hooks; again, this points towards the use of an option type.

However, both of those ignore the more important actually-broken behaviour when you look at the bigger picture of how this function is used during instruction execution. Many of the floating-point instructions also update fflags. But write_fflags gets called before F(rd) is assigned. So `fadd.d x3, x6, x8` on RV32 will fault with an illegal instruction exception, as it should, but already have updated fflags. You could "fix" this by swapping the order, but that is a gross hack. The correct thing to do is to push the actual check that the registers are valid into the guards on the encdec clauses or as the very first thing the execute clauses do, and then these functions can just return the value directly, no Retired or option needed. They should probably still have _asserts_ that the input is sane though to catch any caller that forgot to check the registers.

Reply:
Hi @jrtc27 

I am wondering if there is a better way of achieving the following snippet 

`
overload operator | = {or_vec,or_two_bits,or_bit_and_bool}
val or_two_bits : (bit,bit) -> bool
function or_two_bits (x,y) = {
  bit_to_bool(x) | bit_to_bool(y)
}

val or_bit_and_bool : (bit,bool) -> bool
function or_bit_and_bool (x,y) = {
  bit_to_bool(x) | y
}`

I am trying to check that non of any number of `bit` (be it even or odd is bitone), and I ended up with this but maybe there is a better way of achieving the same thing ?  Would this be accepted in the model ? 

Regards,
Ibrahim

Reply:
That snippet is completely wrong, `bit | bit` is a bit, `bool | bool` is a bool and any other combination is incorrect code, because bits and bools are completely different things.

But I also don't understand what that has to do with what you're saying about "any number of bit", nor what you're actually saying.

Reply:
```
mapping clause encdec =
    F_BIN_RM_TYPE_D(rs2, rs1, rm, rd, FADD_D)
    if (is_RV32D() & ( ~(sys_enable_zfinx()) | (rs1[0] | rs2[0] | rd[0]) == bitzero)) | is_RV64D()
<-> 0b000_0001 @ rs2 @ rs1 @ encdec_rounding_mode (rm) @ rd @ 0b101_0011
    if (is_RV32D() & ( ~(sys_enable_zfinx()) | (rs1[0] | rs2[0] | rd[0]) == bitzero)) | is_RV64D()
```
Hi @jrtc27  I am considering adding the following conditions to each RV32/64D encdec clause, does it looks all right ? I am not sure if this would be considered too repetitive ?

Reply:
That looks correct but yes you should extract the valid register check into a function

Reply:
I was thinking that, but I am not sure that would help much, because either I would create a single function that checks a single register which would be almost the same, or I would create a single function that takes all the registers and return a boolean, but then I would have to create another one for fmadd case and does not feel much better.

Reply:
So either have a `bits(5) -> bool` and a trio of `(bits(5), bits(5)) -> bool`, `(bits(5), bits(5), bits(5)) -> bool` and `(bits(5), bits(5), bits(5), bits(5)) -> bool` functions (you can use an overload for them all) or have a single `forall 'n, 'n > 0. vector('n, dec, bits(5)) -> bool` function.

Reply:
```
val registers_even : forall 'n, 'n > 0. (vector('n, dec, bits(5))) -> bool
function reg_even (registers) = {
  foreach (x from 0 to (length(registers) - 1)) {
    if (registers[x][0] == bitone) then return false
  };
  true
}
```
@jrtc27  I tried this, SAIL happily compiles it, but it result in a linker error ` undefined reference to length_zz5vecz8z5bv5z9`, I will for now pass the length of the vector, and I will report this as a bug to the SAIL compiler, unless you see something wrong with it, PS: I import vector_dec.sail

Reply:
Yeah ok I see the problem, ran into somewhat similar things before.

You don't quite need an actual parameter though, you can use `implicit('n)` so as far as the callers are concerned there's only the vector argument.

Reply:
And don't make the function a registers_even function (unless it's a helper for the helper), pull _all_ the logic out into a function otherwise you'll still need to repeat the `~(zfinx) | rv64 | even(...)` mess everywhere. This way it's just an "are these floating point registers valid operands for a double operation" function.

Reply:
Thanks, implicit appears to have worked.
Yes, I am planning to put it all in a single function, was trying to see how liquid types work with vectors first before placing the rest of the logic in that function..

Reply:
> use `implicit('n)`

Reply:
@jrtc27  Going back to the `proper discriminated union`, if I use option(bits(64)), then I would need to either use Some on all cases where I call X_or_F_D or need to probably check for, and since I already moved the checking to encdec statements, I am not sure this really make sense as it should in theory never happen ? Any suggestion if this is really necessary.

Reply:
Since it should never happen, make it an assert. The reason I talked about using an option type, rather than just mentioning the trap behaviour issue and ignoring the no-longer-relevant types + layering violation issue, was so that you're aware of all the issues with your original approach and thus better informed for any future work.

Reply:
My usage for those dummy values are like this one : 
https://github.com/riscv/sail-riscv/blob/master/model/riscv_insts_fext.sail#L379-L382

Reply:
> here is a layering violation

I think what @jrtc27  means is that you have a return type `... -> option(unit)` (and possibly drop the `escape` effect) and then let the caller decide what to do with the failed call. This is widely used in the RISCV code base, e.g. in https://github.com/riscv/sail-riscv/blob/master/model/riscv_insts_fext.sail#L375-L382 (note that the last `RETIRE_FAIL` there is also a dummy value to keep the type-checker happy). In an even more expressive language, you'd probably use a monad to avoid the inevitable code duplication at call sites, but Sail doesn't have those.

@jrtc27 is that what you have in mind?

Reply:
Ok, I already have an assert statement, but I have to use such dummy values for the type checker to be happy !  I will update the PR shortly for other issues.

Reply:
Which lives in riscv_insts_fext.sail, not riscv_fdext_regs.sail, is under `Execution semantics` and specifically a helper function for execution floating-point load instructions to factor out common code, not a general "read a register" function that can have other uses.

Reply:
Though, whilst the layering is fine there, it should still be an assert in that function, rather than silently becoming a NOP that doesn't increment minstret

Reply:
> Ok, I already have an assert statement, but I have to use such dummy values for the type checker to be happy ! I will update the PR shortly for other issues.

You should not need to; if you do you're almost certainly doing something wrong.

Reply:
> it should still be an assert in that function, 

This is a good place for me to ask something I've long wondered about: what's the official RISCV guideline for using `assert` in code?

Reply:
> You should not need to; if you do you're almost certainly doing something wrong.

What do you mean, even if I have assertion, I would need to return some value for type checker to be happy ?

Reply:
> > You should not need to; if you do you're almost certainly doing something wrong.
> 
> What do you mean, even if I have an assertion, I would need to return some value for type checker to be happy ?

If you have an assertion, that case cannot possibly progress past the assertion, therefore you do not need to handle it.

Reply:
Yes I know, but type checker wont compile it if I dont have a return value of the correct type, I guess thats why they did it in https://github.com/riscv/sail-riscv/blob/master/model/riscv_insts_fext.sail#L379-L382. It would never get there, but we still need it.

Reply:
Why are you trying to add an if for something you just asserted can't happen? That's the whole point, you don't handle the condition, you just assume it can't happen because if it did you've already crashed with an assertion failure.

Reply:
Because it would be an incomplete if clause, and there there would be no return value. The use case I am trying to have an assertion for is when you try to read a 64 bit (X_or_F_D) and zfinx is not enabled.  it should never reach there because of conditions on encdec, but I think it still need to be there.

Reply:
As I said, why are you trying to add an if? It serves no purpose other than to make your life harder. You know it will always (or never, depending which way round you write it) be taken. The whole point of asserts is so that there is no if, and having both an assert and an if for the same condition is a serious code smell.

Reply:
I guess I should just convert the last case of the if clauses to else, and just have the assertions.

Reply:
... yes that's what I've been saying.

Note that there are no "if clause*s*" plural. There's if and there's else; else if is just else (if ....), there is nothing in the grammar for "else if". This is also true in C-based languages.

Reply:
https://github.com/riscv/sail-riscv/blob/01720566318e292df12693bba1fe5f09ea0857df/model/riscv_insts_dext.sail#L946-L947, Is there any reason why haveDExt() is used instead of is_RV32D_or_RV64D in these lines ? This seems slightly inconsistent, I think I would assume the conditions for this one should match feq.d etc.

Reply:
PS: Now I am working based on the assumption that regex `\w,\w` should not return any results in the model, and every time there is a comma in a function declaration, vector declaration, argument passing etc, there should be a space after it. This assumption seems mostly true with the exception riscv_termination_common.sail !

Reply:
PS: Now I am working based on the assumption that regex `\w,\w` should not return any results in the model, and every time there is a comma in a function declaration, vector declaration, argument passing etc, there should be a space after it. This assumption seems mostly true with the exception riscv_termination_common.sail !

Reply:
https://github.com/riscv/sail-riscv/blob/01720566318e292df12693bba1fe5f09ea0857df/model/riscv_insts_cdext.sail#L79-L82 
Similarly, the conditions in this file should have been is_RV32D_or_RV64() instead of redefining it across the file, but I wont fix it in this PR as its not directly related to Zfinx

Reply:
@jrtc27  I think I am done with all the changes you requested now, could you have a look at it, and let me know what you think ?